### ISSUE TITLE
`GestureDetector` `onPan` now triggers via trackpad. changed application behaviour

### ISSUE THREAD

**Issue by ltOgt** [flutter/flutter#107005]
https://github.com/flutter/flutter/pull/89944 has updated `GestureDetector` so that `onPan` and others are now triggered via trackpad gestures.

Previously `onPan...` was triggered only via click-and-drag on trackpads (like with a regular mouse).

This is still the behaviour in web.

____

This changes behaviour of existing apps.

Take the following example, which implements a selection box via click-and-drag.

<details>

 <summary>Minimal Example + Workaround</summary>

```dart
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  Offset? dragStart;
  Offset? dragEnd;

  bool workaround = true;

// This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        floatingActionButton: FloatingActionButton(
          child: Icon(workaround ? Icons.alt_route_rounded : Icons.bug_report),
          onPressed: () => setState(() {
            workaround = !workaround;
          }),
        ),
        body: GestureDetector(
          onPanStart: (details) => setState(() {
            if (workaround && details.kind == PointerDeviceKind.trackpad) {
              return;
            }
            dragStart = details.localPosition;
          }),
          onPanUpdate: (details) => setState(() {
            if (workaround && dragStart == null) {
              return;
            }
            dragEnd = details.localPosition;
          }),
          onPanEnd: (details) => setState(() {
            dragStart = null;
            dragEnd = null;
          }),
          child: Stack(
            children: [
              Container(
                color: Colors.black,
              ),
              if (dragEnd != null)
                Positioned.fromRect(
                  rect: Rect.fromPoints(dragStart!, dragEnd!),
                  child: Container(
                    color: Colors.blue,
                  ),
                ),
            ],
          ),
        ),
      ),
    );
  }
}
```
</details>

Without adding the following workarounds:
```dart
onPanStart: (details) => setState(() {
  if (workaround && details.kind == PointerDeviceKind.trackpad) {
    return;
  }
  dragStart = details.localPosition;
}),
onPanUpdate: (details) => setState(() {
  if (workaround && dragStart == null) {
    return;
  }
  dragEnd = details.localPosition;
}),
```

this drag behaviour is now also triggered by two-finger scroll on the trackpad:

https://user-images.githubusercontent.com/24209580/177008093-02d7960a-9d1c-496a-84dd-913a2cebb155.mov

This happens only on `desktop` (`macos` at least), not in `web`.
On `Web` only click-and-drag triggers `onPan`.

_____

Similarly `Listener` reports trackpad pan on `desktop` via `onPointerPanZoomUpdate` while on the `web` you must use `onPointerSignal` with `if (event is PointerScrollEvent)`.
(This was previously also called for trackpad scroll on macos)

<details>

```dart
// web
onPointerSignal: (event) {
  if (event is PointerScrollEvent) {
    // do stuff
  }
},
// macos
onPointerPanZoomUpdate: (PointerPanZoomUpdateEvent event) {
  // do stuff
},

```

</details>

_____
```
$ flutter doctor -v
[✓] Flutter (Channel master, 3.1.0-0.0.pre.1400, on macOS 12.1 21C52 darwin-arm, locale en-DE)
    • Flutter version 3.1.0-0.0.pre.1400 on channel master at /Users/omni/development/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 7ac9fc5bae (4 days ago), 2022-06-28 08:39:08 -0400
    • Engine revision 6cff45c25c
    • Dart version 2.18.0 (build 2.18.0-232.0.dev)
    • DevTools version 2.14.1

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 13C100
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 12.1 21C52 darwin-arm
    • Chrome (web)    • chrome • web-javascript • Google Chrome 103.0.5060.53

```

**Comment by ltOgt** [comment]
Note that the click-and-drag behaviour is also explicitly mentioned in this [Widget of the Week](https://youtu.be/WhVXkCFPmK4?t=49) video in the `GestureDetector` [doc](https://api.flutter.dev/flutter/widgets/GestureDetector-class.html).
`Watch "pan" updates to make a widget draggable`.

Here is a reproduction of the example in the video:
```dart
import 'dart:math';

import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  double _top = 0, _left = 0;

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Stack(
          children: [
            Positioned(
              top: _top,
              left: _left,
              child: GestureDetector(
                onPanUpdate: (details) {
                  _top = max(0, _top + details.delta.dy);
                  _left = max(0, _left + details.delta.dx);
                  setState(() {});
                },
                child: Container(
                  width: 100,
                  height: 100,
                  color: Colors.red,
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}
```

Here again, the "drag" is now also triggered by trackpad scroll on macos:

https://user-images.githubusercontent.com/24209580/177033736-d9cc322a-e377-40f3-baac-d29071f8b597.mov

I think the change to have trackpad trigger a callback that was previously only triggered by pointer drag may be breaking a few applications in the wild.

**Comment by huycozy** [comment]
@ltOgt thanks for more detailed response, but I still can not reproduce the issue as yours. 

- With `Side 1` case, two-finger scrolling was not triggered on both `Web` and `macos`. Could you provide device info (model, os version)?

    - My device info: Macbook Pro 2020, MacOS Monterey 12.2.1
    - Trackpad dragging setting:
![Screen Shot 2022-07-05 at 11 06 52](https://user-images.githubusercontent.com/104349824/177247479-3db9e1e4-2f37-4fd6-8796-8700caa47328.png)

- With `Side 2`, I see there inconsistent `Expected` result between `Side 1` and `Side 2`:

> Expected: Trackpad scroll on a mac triggers onPan for GestureDetector.

> Expected: onPan for GestureDetector should only be triggered by pointer click-and-drag.

As I understand this after all Sides, you would expect that `onPan for GestureDetector should only be triggered by pointer click-and-drag.`, right?

**Comment by ltOgt** [comment]
- Device Info: macOS Monterey Version 12.1, MacBook Air (m1, 2020)
- Trackpad settings: i dont have the `three finger drag` enabled, which i think is the default.
<img width="377" alt="image" src="https://user-images.githubusercontent.com/24209580/177308969-cf4dacb3-cb6b-4114-a1ba-5ff7801f0a74.png">

When i enable the option like you, the described behaviour is triggered by three-finger-swipe.
If you disable the option like me, the described behaviour is triggered by two-finger-swip.

So i think with this you can reproduce my behaviour, correct?
Good catch, I would not have thought of this option.

____

Yes the `expected` between `side 1` and `side 2` are conflicting.

# Side 2: "breaking change" from old behaviour to new behaviour
The behaviour prior to https://github.com/flutter/flutter/pull/89944 was that only click-and-drag triggered `onPan`.
This is the behaviour that is described in the [Widget of the Week](https://youtu.be/WhVXkCFPmK4?t=49) video in the `GestureDetector` [doc](https://api.flutter.dev/flutter/widgets/GestureDetector-class.html).

Here the problem is that the behaviour of existing applications has been changed by the PR.
This is kind of a breaking change, as `onPan` has now to be explicitly filtered for `details.kind == PointerDeviceKind.trackpad` to restore the behaviour before the PR.

# Side 1: inconsistency in the new behaviour
Accepting that trackpad scroll now triggers `onPan`, it should at least do so consistently on both `desktop` and `web`.

Instead, on `web`, only `click-and-drag` triggers `onPan`.

_____
As mentioned in the first post, this is not only the case for `GestureDetector`, but also for `Listener`:

# Side 2: "breaking change"

`Listener` used to trigger `onPointerSignal` (with `(event is PointerScrollEvent)`) for trackpad scroll.
Now instead `onPointerPanZoomUpdate` is triggered by trackpad scroll.
This is a breaking change.

# Side 1: inconsistency
On `web`, trackpad scroll still triggers `onPointerSignal` (instead of `onPointerPanZoomUpdate`)

**Comment by maheshmnj** [comment]
Hi @ltOgt, Thanks for filing the issue. I can reproduce the issue scrolling with two fingers on a trackpad triggers the pan gesture.  This is reproducible on the latest master channel and not on the stable channel on macOS (silicon) and Linux (intel) desktops.

*In the below video I am scrolling with two fingers on the trackpad*

https://user-images.githubusercontent.com/31410839/177326318-9c08f022-53e2-4301-9295-d423cafb7937.mov

<details>
<summary>flutter doctor -v (mac silicon)</summary>

```
[✓] Flutter (Channel stable, 3.0.4, on macOS 12.4 21F79 darwin-arm, locale en-IN)
    • Flutter version 3.0.4 at /Users/mahesh/Documents/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 85684f9300 (4 days ago), 2022-06-30 13:22:47 -0700
    • Engine revision 6ba2af10bb
    • Dart version 2.17.5
    • DevTools version 2.12.2

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0-rc4)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-32, build-tools 33.0.0-rc4
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] IntelliJ IDEA Community Edition (version 2021.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 61.2.4
    • Dart plugin version 212.5080.8

[✓] VS Code (version 1.67.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.44.0

[✓] Connected device (3 available)
    • sdk gphone arm64 (mobile) • emulator-5554 • android-arm64  • Android 11 (API 30) (emulator)
    • macOS (desktop)           • macos         • darwin-arm64   • macOS 12.4 21F79 darwin-arm
    • Chrome (web)              • chrome        • web-javascript • Google Chrome 103.0.5060.53

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

```
[✓] Flutter (Channel master, 3.1.0-0.0.pre.1494, on macOS 12.4 21F79 darwin-arm, locale en-IN)
    • Flutter version 3.1.0-0.0.pre.1494 on channel master at /Users/mahesh/Documents/flutter_master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 829bddf616 (3 hours ago), 2022-07-04 03:49:08 -0400
    • Engine revision e78ed7f1c9
    • Dart version 2.18.0 (build 2.18.0-251.0.dev)
    • DevTools version 2.15.0

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0-rc4)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-32, build-tools 33.0.0-rc4
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 13C100
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] IntelliJ IDEA Community Edition (version 2021.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 61.2.4
    • Dart plugin version 212.5080.8

[✓] VS Code (version 1.67.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.44.0

[✓] Connected device (3 available)
    • sdk gphone arm64 (mobile) • emulator-5554 • android-arm64  • Android 11 (API 30) (emulator)
    • macOS (desktop)           • macos         • darwin-arm64   • macOS 12.4 21F79 darwin-arm
    • Chrome (web)              • chrome        • web-javascript • Google Chrome 103.0.5060.53

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```
</details>

<details>
<summary>flutter doctor -v (Linux intel)</summary>

```
[✓] Flutter (Channel stable, 3.0.4, on Ubuntu 20.04.4 LTS 5.13.0-48-generic, locale en_GB.UTF-8)
    • Flutter version 3.0.4 at /home/mahesh/Development/stable
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 85684f9300 (5 days ago), 2022-06-30 13:22:47 -0700
    • Engine revision 6ba2af10bb
    • Dart version 2.17.5
    • DevTools version 2.12.2

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /home/mahesh/Android/Sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: /home/mahesh/Downloads/android-studio-2020.3.1.25-linux/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • clang version 10.0.0-4ubuntu1
    • cmake version 3.16.3
    • ninja version 1.10.0
    • pkg-config version 0.29.1

[✓] Android Studio (version 2020.3)
    • Android Studio at /home/mahesh/Downloads/android-studio-2020.3.1.25-linux/android-studio
    • Flutter plugin version 67.1.2
    • Dart plugin version 212.5744
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[✓] VS Code (version 1.68.0)
    • VS Code at /usr/share/code
    • Flutter extension version 3.42.0

[✓] Connected device (2 available)
    • Linux (desktop) • linux  • linux-x64      • Ubuntu 20.04.4 LTS 5.13.0-48-generic
    • Chrome (web)    • chrome • web-javascript • Google Chrome 102.0.5005.115

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```
```
[✓] Flutter (Channel master, 3.1.0-0.0.pre.1396, on Ubuntu 20.04.4 LTS 5.13.0-48-generic, locale en_GB.UTF-8)
    • Flutter version 3.1.0-0.0.pre.1396 on channel master at /home/mahesh/Development/master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 23437d93ca (7 days ago), 2022-06-27 23:39:08 -0400
    • Engine revision 6301c5cb30
    • Dart version 2.18.0 (build 2.18.0-232.0.dev)
    • DevTools version 2.14.1

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /home/mahesh/Android/Sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: /home/mahesh/Downloads/android-studio-2020.3.1.25-linux/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • clang version 10.0.0-4ubuntu1
    • cmake version 3.16.3
    • ninja version 1.10.0
    • pkg-config version 0.29.1

[✓] Android Studio (version 2020.3)
    • Android Studio at /home/mahesh/Downloads/android-studio-2020.3.1.25-linux/android-studio
    • Flutter plugin version 67.1.2
    • Dart plugin version 212.5744
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[✓] VS Code (version 1.68.0)
    • VS Code at /usr/share/code
    • Flutter extension version 3.42.0

[✓] Connected device (2 available)
    • Linux (desktop) • linux  • linux-x64      • Ubuntu 20.04.4 LTS 5.13.0-48-generic
    • Chrome (web)    • chrome • web-javascript • Google Chrome 102.0.5005.115

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```
</details>

**Comment by maheshmnj** [comment]
cc: @moffatman based on your change in https://github.com/flutter/flutter/pull/89944

**Comment by moffatman** [comment]
Hi @ltOgt 

I think I will be the best position to explain as author of #89944. You can find some detail on the design decisions behind the change [here](https://docs.google.com/document/d/1oRvebwjpsC3KlxN1gOYnEdxtNpQDYpPtUFAkmTUe-K8/edit?resourcekey=0-pt4_T7uggSTrsq2gWeGsYQ). 

It's expected that some behaviours have changed, as the new design allows trackpad gesture features which weren't possible before, but anyone who was used to `onPointerSignal` might have to adapt their code. 

I chose this design to match the same approach as when Apple added trackpad support to UIKit for iPads, existing touch gesture recognizers were updated to trigger on trackpad events. The intention is that trackpad gestures will work in as many places as possible by default. For most pan interactions, having them respond to a trackpad gesture fits very well. For example: scrolling up/down or left/right, panning around content, dragging away a pane, swiping something left/right. I agree that if you have a custom interaction built using GestureRecognizer, the behavior has changed. To make it easier, there should probably by a parameter on `GestureDetector` to filter by `deviceKind` rather than doing it in each callback. It would be useful if you had some interaction that wasn't suitable to be triggered through the trackpad. 

Unfortunately, it won't be possible to add the new trackpad behavior to `web`, the APIs provided by browsers are not sufficient.

**Comment by goderbauer** [comment]
@dkwingsmt

**Comment by ltOgt** [comment]
Thanks for the details @moffatman!

I do appreciate the increased support for trackpad gestures, and the point about "work in as many places as possible by default" makes sense.

Browser APIs not exposing the trackpad itself is a bummer, but nothing can be done about that...

The actual migration from `Listener.onPointerSignal` to `Listener.onPointerPanZoomUpdate` is easily done, and adding a check for `details.kind == PointerDeviceKind.trackpad` to `GestureDetector.onPan` is also not hard.

I guess my main worry is about discoverability of this change for the average consumer of these APIs, especially those that have used them in existing apps.

We dont have any deprecation warning for this, since nothing is really deprecated, but the changes can end up breaking existing applications.

Two existing applications of mine suddenly no longer worked as expected (selection box triggered via pan & pan via `onPointerSignal` not being triggered any more).

I only found about these changes via the inline "last change on git" thingy in vscode.

_____

Also the documentation might need to include information on this new behaviour.

`Listener.onPointerPanZoomStart` hints at trackpad, but does not mention that this does not work on web, and instead `onPointerSignal` should be used on web.
```dart
/// Called when a pan/zoom begins such as from a trackpad gesture.
final PointerPanZoomStartEventListener? onPointerPanZoomStart;
```

`GestureDetector.onPan..` talks only about pointers contacting the screen, and not at all about trackpads:
```dart
/// A pointer has contacted the screen with a primary button and has begun to
/// move.
///
/// See also:
///
///  * [kPrimaryButton], the button this callback responds to. 
final GestureDragStartCallback? onPanStart;
```

Lastly, the example shown in the [Widget of the Week](https://youtu.be/WhVXkCFPmK4?t=49) video in the `GestureDetector` [doc](https://api.flutter.dev/flutter/widgets/GestureDetector-class.html) is now triggered by trackpad as well, which is very much unexpected when watching that video as an introduction to `GestureDetector`.

**Comment by dkwingsmt** [comment]
It would be best to provide a migration guide. We might not be able to cover every cases, but at least some common cases, or at least send an alert. Unfortunately I'm not familiar enough with trackpad. @moffatman Do you think you can draft one? It should be something similar to your previous example, but using the template as described [here](https://github.com/flutter/flutter/wiki/Tree-hygiene#3-prepare-your-change).

**Comment by likang** [comment]
Maybe this pull flutter/engine#31593 caused this issue. the `signal_kind` of `FlutterPointerEvent` is no longer assigned to `kFlutterPointerSignalKindScroll` if event was triggered by Trackpad.

In the result, we can not distinguish scroll event from pan event of Trackpad.

**Comment by dkwingsmt** [comment]
@likang Can you further illustrate what problem it brings, better with a demonstration (reproduction step)?

**Comment by likang** [comment]
Sure, actually @ltOgt posted [here](https://github.com/flutter/flutter/issues/107005#issuecomment-1173046562) demonstrating what happened  clearly: after PR  flutter/engine#31593 , scrolling on trackpad will trigger pan event instead of scroll event.

Another change was we can't get `PointerScrollEvent` from `Listnener` any more. here I add a `Listener` as parent widget of `GestureDetector` in @ltOgt 's demo.

```dart
import 'dart:math';

import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  double _top = 0, _left = 0;

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Stack(
          children: [
            Positioned(
              top: _top,
              left: _left,
              child: Listener(
                onPointerSignal: (event) {
                  if (event is PointerScrollEvent) {
                    debugPrint('scrolling');
                  }
                },
                child: GestureDetector(
                  onPanUpdate: (details) {
                    _top = max(0, _top + details.delta.dy);
                    _left = max(0, _left + details.delta.dx);
                    setState(() {});
                  },
                  child: Container(
                    width: 100,
                    height: 100,
                    color: Colors.red,
                  ),
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}
```

Scroll trackpad  using Flutter on stable channel:
1. `scrolling` message will be printed
2. red block will not change it's position

Scroll trackpad using Flutter on beta channel 
1. `scrolling` message will not be printed
2. red block will change it's position

**Comment by moffatman** [comment]
It's true, I guess it's a breaking change. I was waiting for https://github.com/flutter/flutter/commit/7b07b85d919e184ab8deb62141e222b47eb2bcfe to get a version number to write up the migration guide, it would be helpful in cases like this.

**Comment by likang** [comment]
adding `supportedDevices` is a good idea, it helps on some scenarios. 

but what can we do if we want both `pan` and `scroll` event just like before? Image we have a big scrollable canvas and some widgets in it, we want two-fingers pan to `scroll` canvas, and click-and-pan to draw a marquee.

**Comment by moffatman** [comment]
So in that case, you could use `PanGestureRecognizer` with `supportedDevices={PointerDeviceKind.trackpad}` to recognize the trackpad panning. And another `PanGestureRecognizer` with `supportedDevices={PointerDeviceKind.mouse}` to recognize the click-and-drag. The `PointerDeviceKind.trackpad` will only occur if it is a two-finger drag, if the user is using click-and-drag, even if it's with a trackpad, the `deviceKind` will still be `PointerDeviceKind.mouse`. And continue to use `Listener` to recognize traditional mouse wheel scroll.

**Comment by likang** [comment]
OK, sounds it's a little complicated, but I am going to try it out.

**Comment by theniceboy** [comment]
Hi @moffatman, thanks for the explanation. It seems odd to me, though, that this can't be supported on Flutter web. Web apps such as Figma supports trackpad pan/zoom gestures. I also found [this article](https://kenneth.io/post/detecting-multi-touch-trackpad-gestures-in-javascript) about detecting multi-touch trackpad gestures in JavaScript.

It'll be great if this feature is also supported on Flutter web. Could this still be a possibility? Thanks!

**Comment by moffatman** [comment]
@theniceboy I did see that article before and wasn't able to get all the data I needed using the listed techniques. Just tried out Figma though and to my surprise it works quite well at least for pan/zoom, I guess it's possible trackpad scrolling could be done well in flutter web. Upon initial investigation, they are doing some very interesting hacks and guesswork in the event handling, I'll look into it and try to copy their technique.

**Comment by moffatman** [comment]
@theniceboy Yeah, I already knew I can get the invormation needed for `onPointerPanZoomUpdate`, but sadly it is still not possible to know when the gestures start and stop, so can't do `onPointerPanZoomStart` or `onPointerPanZoomEnd`. That means we can't have full continuous gesture support. Pan should work already using the older discrete way (`onPointerScroll`), sadly you can't do any cool gesture interactions like pull-to-refresh this way. Probably the best thing to do now is to make a new type of event (`onPointerScale`) for use only in the web. Going to try that out and see how it feels.

**Comment by theniceboy** [comment]
I see. Thank you for the explanation! Currently `onwheel` provides what I needed for my application, but I see a lot of things can't be done with this call. Just FYI here's the code I'm using:

In `index.html`:
```javascript
document.addEventListener(
	"wheel",
	function (e) {
		e.cancelable && e.preventDefault();
		window.onWebTrackpadPanZoom(e.deltaX, e.deltaY, e.ctrlKey, e.x, e.y);
	},
	{ capture: true, passive: false },
);
```

In `main.dart`:
```dart
js.context["onWebTrackpadPanZoom"] = js.allowInterop((...) {
  ...
});
```

Regarding pull-to-refresh, I feel like most users don't do that with their trackpad, but I'm not a very heavy trackpad user so I'm not sure.

**Comment by gspencergoog** [comment]
It seems like there isn't really much left to do here, since it's "as designed", although the idea to filter on device type in the gesture detector seems like a good one.

Since there's not anything left to do for the original issue (as far as I can tell), we're going to close this.

@moffatman If you could file a feature request for the additional flag on GestureDetector, that would be great.

(Desktop Triage)

**Comment by moffatman** [comment]
@gspencergoog `supportedDevices` for `GestureDetector` landed in #107312 already
